Issue #2504025 by Fabianx, Pol: Add static container factory method for plugins (D8 style)

drupol opened the pull request.

Commit:
Update createInstance() method.

Commit:
Update code.

fabianx-ai commented on src/Plugin/ContainerAwarePluginManager.php:
```diff
@@ -62,10 +63,17 @@ public function createInstance($plugin_id, array $configuration = array()) {
+    $plugin_class = static::getPluginClass($plugin_id, $plugin_definition);
+
+    // If the plugin provides a factory method, pass the container to it.
+    if (is_subclass_of($plugin_class, 'Drupal\service_container\Plugin\ContainerAwarePluginManager')) {
```
This needs to be the Interface that I outlined, not our own.

We are okay to add that interface, too.

fabianx-ai commented on src/Plugin/ContainerAwarePluginManager.php:
```diff
@@ -62,10 +63,17 @@ public function createInstance($plugin_id, array $configuration = array()) {
     $plugin_definition += array(
       'arguments' => array(),
     );
+    $plugin_class = static::getPluginClass($plugin_id, $plugin_definition);
```
This needs to be moved above the "$plugin_definition += array(".

Commit:
Add ContainerFactory and update code based on comments.

fabianx-ai commented on src/Plugin/ContainerAwarePluginManager.php:
```diff
@@ -67,6 +70,12 @@ public function createInstance($plugin_id, array $configuration = array()) {
     array_unshift($plugin_definition['arguments'], $configuration);
 
+    // If the plugin provides a factory method, pass the container to it.
+    if (is_subclass_of($plugin_class, 'Drupal\Core\Plugin\Factory\ContainerFactory')) {
```
That is the wrong class.

Needs to be:

Drupal\Core\Plugin\ContainerFactoryPluginInterface

fabianx-ai replied:
Move this _above_ the $plugin_definition +=

LionsAd commented on src/Plugin/ContainerAwarePluginManager.php:
```diff
@@ -59,6 +60,8 @@ public function hasDefinition($plugin_id) {
    */
   public function createInstance($plugin_id, array $configuration = array()) {
     $plugin_definition_copy = $plugin_definition = $this->getDefinition($plugin_id);
```
Move the copy making below the check for the is_subclass_of.

Commit:
Fix Service Container Annotation Discovery tests.
Provide tests for this issue, fix namespaces, directories and erroneous class names.

drupol commented:
Investigating...

There were 2 errors:
1) Drupal\Tests\service_container\Plugin\ContainerAwarePluginManagerTest::test_createInstance
Drupal\Component\Plugin\Exception\PluginException: Plugin (block) instance class "Mockery_11__Drupal_render_cache_block_RenderCache_Controller_BlockController#0000000034594da8000000002ba5df98" does not exist.
/home/travis/build/LionsAd/service_container/src/Plugin/ContainerAwarePluginManager.php:120
/home/travis/build/LionsAd/service_container/src/Plugin/ContainerAwarePluginManager.php:63
/home/travis/build/LionsAd/service_container/tests/src/Plugin/ContainerAwarePluginManagerTest.php:58
2) Drupal\Tests\service_container\Plugin\ContainerAwarePluginManagerTest::test_getInstance
Drupal\Component\Plugin\Exception\PluginException: Plugin (block) instance class "Mockery_11__Drupal_render_cache_block_RenderCache_Controller_BlockController#0000000034594dca000000002ba5df98" does not exist.
/home/travis/build/LionsAd/service_container/src/Plugin/ContainerAwarePluginManager.php:120
/home/travis/build/LionsAd/service_container/src/Plugin/ContainerAwarePluginManager.php:63
/home/travis/build/LionsAd/service_container/src/Plugin/ContainerAwarePluginManager.php:89
/home/travis/build/LionsAd/service_container/tests/src/Plugin/ContainerAwarePluginManagerTest.php:68

fabianx-ai commented on modules/providers/service_container_annotation_discovery/tests/modules/service_container_annotation_discovery_test/src/Plugin/Plugin3/Plugin3A/Plugin3A.php:
```diff
@@ -0,0 +1,27 @@
+   * @inheritdoc
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return 'Hello World!';
```
No, that is wrong.

This needs to be:

return new static($configuration, $plugin_id, $plugin_definition, 'Hello World');

then add a __construct here, that populates $this->data and add a getData() method.

drupol replied:
Actually, I wrote that because I didn't know how to test it correctly.

fabianx-ai commented on src/Plugin/ContainerAwarePluginManager.php:
```diff
@@ -59,6 +60,13 @@ public function hasDefinition($plugin_id) {
+
+    // If the plugin provides a factory method, pass the container to it.
+    if (is_subclass_of($plugin_class, 'Drupal\Core\Plugin\ContainerFactoryPluginInterface')) {
+      return $plugin_class::create(\Drupal::getContainer(), $configuration, $plugin_id, $plugin_definition);
```
$this->container instead of \Drupal

Commit:
Update based on comments.

fabianx-ai committed:
Fix test by using get_class().

drupol committed:
Merge pull request #1 from LionsAd/pr--72

Fix test by using get_class().

Commit:
Use yml file for defining the Services and Parameters.

Commit:
This has nothing to do here but in service_container_block (hidden for now) submodule.

Commit:
Add defensive condition to the right place.

Commit:
Revert "Use yml file for defining the Services and Parameters."

This reverts commit cba75f44fd252b9821166f313032c404c2249997.

Commit:
Revert "This has nothing to do here but in service_container_block (hidden for now) submodule."

This reverts commit e0e4cf3cbb1b46baa6c144e90d0244f012de0c31.

Commit:
Revert "Add defensive condition to the right place."

This reverts commit c72fe9700180ff860f0d1c42fc5b07a00721b5a6.

Commit:
Fix dependencies, only depend on service_container.

fabianx-ai commented:
**RTBM**

fabianx-ai commented:
Just waiting for tests and coverage report.

Commit:
Fix test by fixing the test module's dependency.
